src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs

```diff
@@ -0,0 +1,945 @@
+            }
+            else if (attribute.IsTargetAttribute(this, AttributeDescription.MaybeNullAttribute))
+            {
+                arguments.GetOrCreateData<ReturnTypeWellKnownAttributeData>().HasMaybeNullAttribute = true;
+            }
+            else if (attribute.IsTargetAttribute(this, AttributeDescription.NotNullAttribute))
+            {
+                arguments.GetOrCreateData<ReturnTypeWellKnownAttributeData>().HasNotNullAttribute = true;
+            }
+            else if (attribute.IsTargetAttribute(this, AttributeDescription.NotNullIfNotNullAttribute))
+            {
+                arguments.GetOrCreateData<ReturnTypeWellKnownAttributeData>().AddNotNullIfParameterNotNull(attribute.DecodeNotNullIfNotNullAttribute());
+            }
+        }
+
+        private void DecodeDllImportAttribute(ref DecodeWellKnownAttributeArguments<AttributeSyntax, CSharpAttributeData, AttributeLocation> arguments)
+        {
+            Debug.Assert((object)arguments.AttributeSyntaxOpt != null);
+
+            var attribute = arguments.Attribute;
+            Debug.Assert(!attribute.HasErrors);
+            bool hasErrors = false;
+
+            if (!this.IsExtern || !this.IsStatic)
+            {
+                arguments.Diagnostics.Add(ErrorCode.ERR_DllImportOnInvalidMethod, arguments.AttributeSyntaxOpt.Name.Location);
+                hasErrors = true;
+            }
+
+            if (this.IsGenericMethod || ContainingType?.IsGenericType == true)
+            {
+                arguments.Diagnostics.Add(ErrorCode.ERR_DllImportOnGenericMethod, arguments.AttributeSyntaxOpt.Name.Location);
+                hasErrors = true;
+            }
+
+            string moduleName = attribute.GetConstructorArgument<string>(0, SpecialType.System_String);
+            if (!MetadataHelpers.IsValidMetadataIdentifier(moduleName))
+            {
+                // Dev10 reports CS0647: "Error emitting attribute ..."
+                CSharpSyntaxNode attributeArgumentSyntax = attribute.GetAttributeArgumentSyntax(0, arguments.AttributeSyntaxOpt);
+                arguments.Diagnostics.Add(ErrorCode.ERR_InvalidAttributeArgument, attributeArgumentSyntax.Location, arguments.AttributeSyntaxOpt.GetErrorDisplayName());
+                hasErrors = true;
+                moduleName = null;
+            }
+
+            // Default value of charset is inherited from the module (only if specified).
+            // This might be different from ContainingType.DefaultMarshallingCharSet. If the charset is not specified on module
+            // ContainingType.DefaultMarshallingCharSet would be Ansi (the class is emitted with "Ansi" charset metadata flag) 
+            // while the charset in P/Invoke metadata should be "None".
+            CharSet charSet = this.GetEffectiveDefaultMarshallingCharSet() ?? Cci.Constants.CharSet_None;
+
+            string importName = Name;
```

Comment:
>string importName = Name; [](start = 12, length = 25)

I am not sure what is the motivation behind the change here and below.

Reply:
This makes it so we default the import name to the name of the local function in source instead of defaulting it to the name of the synthesized method emitted for the local function.

Without this change it's necessary to pass an explicit EntryPoint to the DllImportAttribute for a 'static extern' local function. It could maybe also be done by adding an ImportName or similar property to source methods which has synthesized methods delegate to the base method where appropriate, but this seemed like a simpler way to do it.

---
In reply to: [354038272](https://github.com/dotnet/roslyn/pull/40011#discussion_r354038272) [](ancestors = 354038272)

Reply:
> Without this change it's necessary to pass an explicit EntryPoint to the DllImportAttribute for a 'static extern' local function.

Instead of making changes in multiple places, consider adjusting value passed to SetDllImport at the call-site below

---
In reply to: [354065673](https://github.com/dotnet/roslyn/pull/40011#discussion_r354065673) [](ancestors = 354065673,354038272)